src/NHibernate/Type/CollectionType.cs

```diff
@@ -369,19 +370,36 @@ public virtual object InstantiateResult(object original)
 			return result;
 		}
 
-		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache,
-											  ISessionImplementor session)
+		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache, ISessionImplementor session)
 		{
-			// TODO: does not work for EntityMode.DOM4J yet!
-			object result = target;
-			Clear(result);
+			var elemType = GetElementType(session.Factory);
+			var resultPc = target as IPersistentCollection;
+			var originalPc = original as IPersistentCollection;
+			var clearTargetsDirtyFlag = false;
+			if (resultPc != null)
+			{
+				if (originalPc == null)
+				{
+					if (!resultPc.IsDirty && AreCollectionElementsEqual((IEnumerable)original, (IEnumerable)target))
+					{
+						clearTargetsDirtyFlag = true;
+					}
+				}
+				else
+				{
+					if (!originalPc.IsDirty)
+					{
+						clearTargetsDirtyFlag = true;
+					}
+				}
+			}
 
+			Clear(target);
 			// copy elements into newly empty target collection
-			IType elemType = GetElementType(session.Factory);
-			IEnumerable iter = (IEnumerable)original;
-			foreach (object obj in iter)
+			var iter = (IEnumerable)original;
```

Comment:
Please move up this line for avoiding casting it too when calling `AreCollectionElementsEqual`.